app/services/articles/feed.rb

```diff
@@ -36,12 +39,60 @@ def default_home_feed_and_featured_story(user_signed_in: false)
           limited_column_select.order("published_at DESC").limit(rand(15..80))
         hot_stories = hot_stories.to_a + new_stories.to_a
       end
+      hot_stories = rank_articles(hot_stories) if @user
       [featured_story, hot_stories]
     end
 
     def default_home_feed(user_signed_in: false)
       _featured_story, stories = default_home_feed_and_featured_story(user_signed_in: user_signed_in)
       stories
     end
+
+    def rank_articles(articles)
+      ranked_articles = articles.map do |article|
+        article_points = 0
+        article_points += score_followed_user(article)
+        article_points += score_followed_organization(article)
+        article_points += score_followed_tags(article)
+        article_points += score_randomness
+        article_points += score_language(article)
+        article_points += score_experience_level(article)
+        RankedArticle.new(article, article_points)
+      end
+      ranked_articles.sort { |a, b| b.score <=> a.score }.map(&:article)
```

Comment:
Note that there's also the option of performing the reverse sort with `ranked_articles.sort_by { |a| -a.score }`. One reason I'm mentioning this is because `sort_by` is generally faster than `sort` and while I'm not a big fan of micro-optimizations, we're talking about the article feed and this code will probably get exercised quite a bit. 

Assuming this really will be a hot path, how about something like the following, which would also completely eliminate the `RankedArticle` struct so we don't need to allocate objects that will immediately need to be garbage collected (untested, but you get the idea):

```ruby
ranked_articles = articles.each_with_object({}) do |article, result|
  article_points = 0
  article_points += score_followed_user(article)
  article_points += score_followed_organization(article)
  article_points += score_followed_tags(article)
  article_points += score_randomness
  article_points += score_language(article)
  article_points += score_experience_level(article)
  result[article_points] = article
end

ranked_articles.sort_by { |k, _| -k }.map(&:last)
```

Essentially: create a hash of `score => article`, then sort by key (score) and extract the values (`Hash#sort_by` returns an array of 2-element arrays).

Reply:
Fantastic suggestions @citizen428! 

Regarding the hash: one potential problem is that articles could very possible have the same ranking score, so storing them in a hash with the key being score could result in some article getting dropped from ranking if their key is overwritten. I suppose I could have a hash with keys of `article.id`, and values of `score`!

(And you just keep on suggesting these optimizations!)

Reply:
Came here to add that, but you already implemented the correct fix. 👍